corgidrp/calibrate_kgain.py

```diff
@@ -0,0 +1,1099 @@
+        raise TypeError('rn_bins1 is not a number')
+    if not isinstance(master_files['rn_bins2'], (float, int)):
+        raise TypeError('rn_bins2 is not a number')
+    if not isinstance(master_files['max_DN_val'], (float, int)):
+        raise TypeError('max_DN_val is not a number')
+    if not isinstance(master_files['signal_bins_N'], (float, int)):
+        raise TypeError('signal_bins_N is not a number')
+    if not isinstance(master_files['logplot1'], (float, int)):
+        raise TypeError('logplot1 is not a number')
+    if not isinstance(master_files['logplot2'], (float, int)):
+        raise TypeError('logplot2 is not a number')
+    if not isinstance(master_files['logplot3'], (float, int)):
+        raise TypeError('logplot3 is not a number')
+    
+    # check parameters
+    if type(stack_arr) != np.ndarray:
+        raise TypeError('stack_arr must be an ndarray.')
+    if np.ndim(stack_arr) != 4:
+        raise CalKgainException('stack_arr must be 4-D (i.e., a stack of '
+                '3-D sub-stacks)')
+    if len(stack_arr) <= 10:
+        raise CalKgainException('Number of sub-stacks in stack_arr must '
+                'be more than 10.')
+    if len(stack_arr) < 20 :
+        warnings.warn('Number of sub-stacks is less than 20, '
+        'which is the recommended minimum number for a good fit ')
+    for i in range(len(stack_arr)):
+        check.threeD_array(stack_arr[i], 'stack_arr['+str(i)+']', TypeError)
+        if len(stack_arr[i]) < 5:
+            raise CalKgainException('A sub-stack was found with less than 5 '
+            'frames, which is the required minimum number per sub-stack')
+        if i > 0:
+            if np.shape(stack_arr[i-1]) != np.shape(stack_arr[i]):
+                raise CalKgainException('All sub-stacks must have the '
+                            'same number of frames and frame shape.')
+    if type(stack_arr2) != np.ndarray:
+        raise TypeError('stack_arr2 must be an ndarray.')
+    if np.ndim(stack_arr2) != 3:
+        raise CalKgainException('stack_arr must be 3-D (i.e., a stack of '
+                '2-D sub-stacks')
+    if len(stack_arr2) < 30:
+        raise CalKgainException('Number of sub-stacks in stack_arr2 must '
+                'be equal to or greater than 30.')
+    check.real_positive_scalar(emgain, 'emgain', TypeError)
+    if emgain < 1:
+        raise CalKgainException('emgain must be >= 1.')
+    check.positive_scalar_integer(min_val, 'min_val', TypeError)
+    check.positive_scalar_integer(max_val, 'max_val', TypeError)
+    if min_val >= max_val:
+        raise CalKgainException('max_val must be greater than min_val')
+    check.positive_scalar_integer(binwidth, 'binwidth', TypeError)
+    if binwidth < 10:
+        raise CalKgainException('binwidth must be >= 10.')
+    if binwidth > 800:
+        raise CalKgainException('binwidth must be < 800.')
+    
+    ################### function defs #####################
```

Comment:
@maxwellmb can you clarify for me what you mean? (these functions are in the file, defined before calibrate_kgain. I do not think we want to add sub-functions inside calibrate_kgain. Adding them to mock.py or elsewhere is not a good idea since they are only used here.)

Reply:
they're currently defined in the `calibrate_kgain` function which makes the function just overall long and harder to read/follow. I was just suggesting that they get defined earlier in the file begfore the `calibrate_kgain` function. This is mostly just to help with code readability.

Reply:
@maxwellmb Oh! That's how it is now -and might have been for a couple of days before I read this conversation-. I'm going to close the convo, unless you want to change it.

Reply:
Done